### DISCUSSION TITLE
How do I acquire interactive user input while quarto render is running?

### DISCUSSION THREAD

**Discussion by harrelfe** [quarto-dev/quarto-cli#6076] [Q&A] [answered]
### Description

I want to interactively provide a password so that R code can unencrypt a file inside a `Quarto` script, so that the password does not need to be stored in the script.   [This](https://xd-deng.com/render_html/keyboard_enter_in_non_interactive_mode.html) works when running plain R scripts. How do I do this for `Quarto` processing, especially when **not** using `RStudio`?

**Comment by cderv** [comment, marked as answer]
You can't use any interactive R code for R chunk executed non-interactively by Quarto. 

I think you can built though some wrapper tooling to ask for your input and then pass it to Quarto or to R directly. 

I see though several solution to pass some information in a "hidden way":

Regarding R way, it would be quite similar to some solutions to secure password to database: https://solutions.posit.co/connections/db/best-practices/managing-credentials/

* Use **keyring** R package to access from R some secrets in some credentials store: https://solutions.posit.co/connections/db/best-practices/managing-credentials/#encrypt-credentials-with-keyring
   Your user would have to setup the keyring and your R code would read inside. 

* Use **config** R to leverage config file : https://solutions.posit.co/connections/db/best-practices/managing-credentials/#stored-in-a-file-with-config
   Your user would have to configure their config file locally, and your R code would read into this local file. This config file could be also reading environment variables as this is a way to have also user-input for a code. 

* Using environment variable is also an option of its own: https://solutions.posit.co/connections/db/best-practices/managing-credentials/#use-environment-variables
   Your user would need to setup and environment variable before running the rendering for R code to access the value


Regarding Quarto itself for non-R specific way 

* Environment is a good way to secure some values: https://quarto.org/docs/projects/environment.html
   There is a mechanism to even make some variables [required](https://quarto.org/docs/projects/environment.html#required-variables) and to make sure users are creating the `.local` version of the file 
   These env vars can then be accessed by any computation code. 


Regarding interactivity, you could create wrapper function in R to help setup those values and then call `quarto render` for the user. 

Hope it helps by giving you some ideas.

**Reply by harrelfe** [reply]
This is **incredibly** helpful.  Thank you!

**Comment by higgi13425** [comment]
Hi @cderv - I am trying to do something similar, but using Keyring. This works when I run the chunks, but when I attempt to render the Quarto doc, it opens an interactive window for me to enter the password, which seems puzzling.

I set the keyring password interactively, then run the following 2 chunks:

{r setup}
#| include: false
library(tidyverse)
library(REDCapTidieR)
library(labelled)
library(keyring)

redcap_uri <- "https://redcapproduction.umms.med.umich.edu/redcap_v14.0.15/API/"
biofire_api_key <- keyring::key_get(service = "biofire_api_key")
data <- read_redcap(redcap_uri, biofire_api_key)

If I just run the chunks, everything works.
But when I try to render, Quarto is looking for an interactive password.

Is there something obvious I am doing wrong?

thanks, Peter

**Comment by higgi13425** [comment]
@cderv, following up on the keyring issue.

Alternatively, if I define the api_key in my .Rprofile, rendering works fine. 
It seems that I am doing something wrong with keyring??

**Comment by higgi13425** [comment]
@cderv, note that I am still seeking a solution that will allow the following
A quarto document/dashboard that can use a secret password to access a REDCap API. pull data, generate a document, and publish once weekly via Github Actions to the web (cron job) without publicly revealing the API key (secret). https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions 
I think I am close, but not quite there.
Frank and the millions of other folks using REDCap 
https://www.project-redcap.org/
may also want a similar problem solved.
- Peter
